contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,205 @@
+        return _grantRoles(resource, roleBitmap, account);
+    }
+
+
+    /**
+     * @dev Revokes roles from `account`.
+     *
+     * The caller must have all the necessary admin roles for the roles being revoked.
+     *
+     * @param resource The resource to revoke roles within.
+     * @param roleBitmap The roles bitmap to revoke.
+     * @param account The account to revoke roles from.
+     * @return `true` if the roles were revoked, `false` otherwise.
+     */
+    function revokeRoles(bytes32 resource, uint256 roleBitmap, address account) public virtual canGrantRoles(resource, roleBitmap) returns (bool) {
+        return _revokeRoles(resource, roleBitmap, account);
+    }
+
+    // Internal functions
+
+    /**
+     * @dev Reverts if `account` does not have all the given roles.
+     */
+    function _checkRoles(bytes32 resource, uint256 roleBitmap, address account) internal view virtual {
+        if (!hasRoles(resource, roleBitmap, account)) {
+            revert EACUnauthorizedAccountRoles(resource, roleBitmap, account);
+        }
+    }
+
+    /**
+     * @dev Reverts if `account` does not have the admin roles for all the given roles.
+     */
+    function _checkCanGrantRoles(bytes32 resource, uint256 roleBitmap, address account) internal view virtual {
+        uint256 adminRoles = roleBitmap & 0xffffffffffffffffffffffffffffffff00000000000000000000000000000000;
+        if (adminRoles == 0 || !hasRoles(resource, adminRoles, account)) {
+            revert EACUnauthorizedAccountAdminRoles(resource, roleBitmap, account);
+        }
+    }
+
+    /**
+     * @dev Copies all roles from `srcAccount` to `dstAccount` within the same resource.
+     *
+     * @param resource The resource to copy roles within.
+     * @param srcAccount The account to copy roles from.
+     * @param dstAccount The account to copy roles to.
+     */
+    function _copyRoles(bytes32 resource, address srcAccount, address dstAccount) internal virtual {
+        uint256 srcRoles = _roles[resource][srcAccount];
+        _roles[resource][dstAccount] |= srcRoles;
+        emit EACRolesCopied(resource, resource, srcAccount, dstAccount, srcRoles);
+    }
+
+    /**
+     * @dev Grants multiple roles to `account`.
+     */
+    function _grantRoles(bytes32 resource, uint256 roleBitmap, address account) internal virtual returns (bool){
+        uint256 currentRoles = _roles[resource][account];
+        uint256 updatedRoles = currentRoles | roleBitmap;
+
+        if (currentRoles != updatedRoles) {
+            _roles[resource][account] = updatedRoles;
+            emit EACRolesGranted(resource, roleBitmap, account, _msgSender());
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * @dev Attempts to revoke roles from `account` and returns a boolean indicating if roles were revoked.
+     */
+    function _revokeRoles(bytes32 resource, uint256 roleBitmap, address account) internal virtual returns (bool) {
```

Comment:
As above - unless we also want to special case allowing a user to revoke their own roles without admin power.

Reply:
This would be the same as the `renounceRoles` method we had but then removed. Am going to leave it like this for now.